src/pool/srv_pool.c

```diff
@@ -4566,10 +4566,19 @@ ds_pool_evict_handler(crt_rpc_t *rpc)
 
 	ABT_rwlock_wrlock(svc->ps_lock);
 
-	rc = find_hdls_to_evict(&tx, svc, &hdl_uuids, &hdl_uuids_size,
-				&n_hdl_uuids);
-	if (rc != 0)
-		D_GOTO(out_lock, rc);
+	/*
+	 * If a subset of handles is specified use them instead of iterating
+	 * through all handles for the pool uuid
+	 */
+	if (in->pvi_hdls.ca_arrays) {
```

Comment:
Should the code iterate through all of the (official, valid) pool handle UUIDs and use any provided pvi_hdls as a filter on that list? Just in case some or all of pvi_hdls are invalid?

Reply:
That is probably a cleaner way to address it but is computationally more work. I guess the question is what happens if you try to do a rdb_tx_delete on a  key that doesn't exist. Will it turn into a noop or will it return an error. If it just ignores the request then its less work io server side to just assume the list is canonical and correct and in the event a handle somehow gets closed but the agent wasn't notified and it asks to clean that handle up we end up just passing a little bit of data that is not used. However if we are going to get an error on the disconnect that will propagate all the way back up the stack then we will have to do it the way you suggest.

Reply:
even before the rdb_tx_delete() in pool_disconnect_hdls() watch for the nhandles -= n_hdl_uuids statement. If the n_hdl_uuids provided in the RPC and passed as input to this function exceeds the number of handles as understood in the pool rdb, then there will be problems. It will try to write back this erroneous nhandles value into the rdb after the rdb_tx_delete() calls are done.

as for rdb_tx_delete() specifying a nonexistent key it is not going to return an error. And it is not going to be a noop. It will instead append an entry into the raft log that will be sent to all of the pool service replicas. Ultimately a replica receiving this raft "appendentries" RPC will flow into the rdb_raft_log_offer()/rdb_raft_log_offer_single()/rdb_tx_apply()/etc. flow, rdb_tx_apply_delete(). Eventually there may be a DER_NONEXIST (I'm speculating) returned by vos_obj_punch() that could result in the log entry being discarded. It seems a little risky as an initial solution here, in the sense that incorrect data is distributed across the service, and once this happens there is no error "feedback" that would be returnable to the client. The raft protocol and rdb itself "should"(?) be OK (if all replicas uniformly discard the entry).  I suppose an experiment could be tried to confirm that the PS replica ioservers may generate the D_ERROR "failed to apply entry" in rdb_raft_log_offer_single().

Maybe a preliminary landing could do the more computationally expensive filtering, and then separately an optimization could be considered - i.e., the above distributed dissemination could be entertained, but would probably require some additional testing to verify the integrity of the service under those kind of conditions.

One idea is to land a safe / correct implementation now and consider an optimization as a separate step.

Reply:
> 
> Maybe a preliminary landing could do the more computationally expensive filtering, and then separately an optimization could be considered - i.e., the above distributed dissemination could be entertained, but would probably require some additional testing to verify the integrity of the service under those kind of conditions.

This does seem like the most prudent approach. Get it right first, then make it faster if necessary. IMO this is not likely to be an area of concern for performance, and spending some extra cycles to avoid generating bogus log entries is a good tradeoff.

Reply:
That is a good catch. I hadn't considered the case where we try to remove 5 from the count of the handles in rdb but only 4 are valid. I'll work on passing the list in as a filter into the filter function instead.